aten/src/ATen/native/cuda/FusedAdamKernel.cu

```diff
@@ -0,0 +1,228 @@
+    const double weight_decay,
+    const double eps,
+    const bool maximize,
+    const bool amsgrad,
+    const float* grad_scale_ptr,
+    const float* found_inf_ptr
+) {
+#pragma unroll
+    for (int ii = 0; ii < kILP; ii++) {
+        // Load values.
+        opmath_t param = static_cast<opmath_t>(r_args[kParamIdx][ii]);
+        opmath_t grad = static_cast<opmath_t>(r_args[kGradIdx][ii]);
+        if (grad_scale_ptr) {
+            grad /= (static_cast<double>(*grad_scale_ptr));
+        }
+        const opmath_t grad_to_store = grad;
+        if (maximize) {
+            grad = -grad;
+        }
+        opmath_t exp_avg = static_cast<opmath_t>(r_args[kExpAvgIdx][ii]);
+        opmath_t exp_avg_sq = static_cast<opmath_t>(r_args[kExpAvgSqIdx][ii]);
+        opmath_t max_exp_avg_sq;
+        if (amsgrad) {
+            max_exp_avg_sq = static_cast<opmath_t>(r_args[kMaxExpAvgSqIdx][ii]);
+        }
+
+        // Update param, grad, 1st and 2nd order momentum.
+        if (weight_decay != 0) {
+            grad += param * weight_decay;
+        }
+        exp_avg = beta1 * exp_avg + (1 - beta1) * grad;
+        exp_avg_sq = beta2 * exp_avg_sq + (1 - beta2) * grad * grad;
+
+        if (amsgrad) {
+            max_exp_avg_sq = ::max(max_exp_avg_sq, exp_avg_sq);
+        }
+
+        const opmath_t bias_correction1 = 1 - ::pow(beta1, *step_count);
+        const opmath_t bias_correction2 = 1 - ::pow(beta2, *step_count);
+
+        const opmath_t step_size = lr / bias_correction1;
+
+        const opmath_t bias_correction2_sqrt = ::sqrt(bias_correction2);
+
+        opmath_t denom;
+        if (amsgrad) {
+            denom = (::sqrt(max_exp_avg_sq) / bias_correction2_sqrt) + eps;
+        } else {
+            denom = (::sqrt(exp_avg_sq) / bias_correction2_sqrt) + eps;
+        }
+
+        param -= step_size * exp_avg / denom;
+
+        // Store results.
+        r_args[kParamIdx][ii] = param;
+        r_args[kGradIdx][ii] = grad_to_store;
+        r_args[kExpAvgIdx][ii] = exp_avg;
+        r_args[kExpAvgSqIdx][ii] = exp_avg_sq;
+        if (amsgrad) {
+            r_args[kMaxExpAvgSqIdx][ii] = max_exp_avg_sq;
+        }
+    }
+}
+
+// [note: Conditional Gradient Store when `optimizer.step` is called by GradScaler]
+// When a user is training their model(s) with an FP16 AMP recipe,
+// parameter updates are done via `grad_scaler.step(optimizer)` instead of `optimizer.step()`.
+// For most optimizers, GradScaler unscales gradients on behalf of those optimizers.
+// Also, before `.step`, it makes sure that all the gradients involved are finite, which incurs a device sync.
+// On the other hand, fused optimizers set their member variable of `_step_supports_amp_scaling` to `True`
+// in order to remove the device sync above. This means that fused optimizers have to have
+// their CUDA kernels (a) unscale gradients and (b) skip parameter updates accordingly.
+// To be functionally on par with `torch.optim` optimizers and `_multi_tensor` ones,
```

Comment:
this is confusing, because currently the interface to optimizers is constructing via torch.optim and setting a flag whether to use mta. In this sense, regular and mta optimizers are functionally equivalent (and mta optimizers typically don't support additional functionality like scaling)

Reply:
um sorry I don't understand. probably due to my bad comment.
What I meant to say is, when `grad_scaler.step(optimizer)` is called and this optimizer is Adam or foreach Adam, the grad scaler sees and unscales the gradient values to call `optimizer.step` (= _single_adam or _multi_tensor_adam function).
On the other hand, if it's fused Adam which sets `_step_supports_amp_scaling` to `True`, `optimizer.step` is called notwithstanding gradients' sanity, so _fused_adam (the functional form) and the cuda kernel need to skip the param update itself and take care of step counts.